v-next/hardhat/src/internal/builtin-plugins/coverage/hook-handlers/solidity.ts

```diff
@@ -0,0 +1,93 @@
+import type { SolidityHooks } from "../../../../types/hooks.js";
+
+import path from "node:path";
+
+import { addStatementCoverageInstrumentation } from "@ignored/edr-optimism";
+import { HardhatError } from "@nomicfoundation/hardhat-errors";
+import { ensureError } from "@nomicfoundation/hardhat-utils/error";
+import { readUtf8File } from "@nomicfoundation/hardhat-utils/fs";
+
+import { unsafelyCastAsHardhatRuntimeEnvironmentImplementation } from "../helpers.js";
+
+const COVERAGE_LIBRARY_PATH =
+  "__hardhat_coverage_library_a3e9cfe2-41b4-4a1f-ad9e-ac62dd82979e.sol";
+
+export default async (): Promise<Partial<SolidityHooks>> => ({
+  preprocessProjectFileBeforeBuilding: async (
+    context,
+    sourceName,
+    fileContent,
+    solcVersion,
+    next,
+  ) => {
+    if (context.globalOptions.coverage) {
+      try {
+        // TODO: Pass the COVERAGE_LIBRARY_IMPORT_PATH as the next argument here
+        // once the EDR supports it.
+        const { source, metadata } = addStatementCoverageInstrumentation(
+          fileContent,
+          sourceName,
+          solcVersion,
+          COVERAGE_LIBRARY_PATH,
+        );
+        // NOTE: We need to cast the hre to the internal HardhatRuntimeEnvironmentImplementation
+        // because the coverage manager (hre._coverage) is not exposed via the public interface
+        const hreImplementation =
+          unsafelyCastAsHardhatRuntimeEnvironmentImplementation(context);
```

Comment:
This part is slightly more confused than needed, as the context is not an HRE.

```
assertHardhatInvariant("_coverage" in context && context._coverage instanceof CoverageManager, "Expected _coverage to be defined in the HookContext, as it's should be defined in the HRE");
```

would be simpler to understand.